File: src/components/graph/styling.rs
use crate::types::PortType;
use commonwl::{
    OneOrMany,
    inputs::{InputSchema, InputType},
    outputs::{CommandOutputParameterType, CommandOutputSchema, CommandOutputType},
    types::CWLType,
};

fn primary_cwl_type(port_type: &PortType) -> Option<CWLType> {
    match port_type {
        PortType::Input(one_or_many) => {
            let iter: Box<dyn Iterator<Item = &InputType>> = match one_or_many {
                OneOrMany::One(t) => Box::new(std::iter::once(t)),
                OneOrMany::Many(v) => Box::new(v.iter()),
            };
            iter.filter_map(|t| match t {
                InputType::CWLType(cwl) => Some(*cwl),
                InputType::String(_) => Some(CWLType::String),
                InputType::InputSchema(_) => None,
            })
            .find(|t| !matches!(t, CWLType::Null))
            .or(Some(CWLType::Null))
        }

        PortType::Output(copt) => match copt {
            CommandOutputParameterType::Stdout | CommandOutputParameterType::Stderr => None, // handled separately below
            CommandOutputParameterType::CommandOutputType(one_or_many) => {
                let iter: Box<dyn Iterator<Item = &CommandOutputType>> = match one_or_many {
                    OneOrMany::One(t) => Box::new(std::iter::once(t)),
                    OneOrMany::Many(v) => Box::new(v.iter()),
                };
                iter.filter_map(|t| match t {
                    CommandOutputType::CWLType(cwl) => Some(*cwl),
                    CommandOutputType::String(_) => Some(CWLType::String),
                    CommandOutputType::CommandOutputSchema(_) => None,
                })
                .find(|t| !matches!(t, CWLType::Null))
                .or(Some(CWLType::Null))
            }
        },
    }
}
fn is_array(port_type: &PortType) -> bool {
    match port_type {
        PortType::Input(OneOrMany::One(InputType::InputSchema(s))) => {
            matches!(**s, InputSchema::Array(_))
        }
        PortType::Output(CommandOutputParameterType::CommandOutputType(OneOrMany::One(
            CommandOutputType::CommandOutputSchema(s),
        ))) => matches!(**s, CommandOutputSchema::Array(_)),
        _ => false,
    }
}

fn is_optional(port_type: &PortType) -> bool {
    match port_type {
        PortType::Input(OneOrMany::Many(v)) => v
            .iter()
            .any(|t| matches!(t, InputType::CWLType(CWLType::Null))),
        PortType::Output(CommandOutputParameterType::CommandOutputType(OneOrMany::Many(v))) => v
            .iter()
            .any(|t| matches!(t, CommandOutputType::CWLType(CWLType::Null))),
        _ => false,
    }
}

fn cwl_type_to_stroke(t: &CWLType) -> &'static str {
    match t {
        CWLType::String => "stroke-red-400",
        CWLType::File => "stroke-green-400",
        CWLType::Directory => "stroke-blue-400",
        CWLType::Boolean => "stroke-yellow-400",
        CWLType::Double => "stroke-purple-400",
        CWLType::Float => "stroke-pink-400",
        CWLType::Long => "stroke-cyan-400",
        CWLType::Int => "stroke-teal-400",
        CWLType::Null | CWLType::Any => "stroke-slate-400",
    }
}

fn cwl_type_to_bg(t: &CWLType) -> &'static str {
    match t {
        CWLType::File => "bg-green-400",
        CWLType::Directory => "bg-blue-400",
        CWLType::String => "bg-red-400",
        CWLType::Boolean => "bg-yellow-400",
        CWLType::Double => "bg-purple-400",
        CWLType::Float => "bg-pink-400",
        CWLType::Long => "bg-cyan-400",
        CWLType::Int => "bg-teal-400",
        CWLType::Null | CWLType::Any => "bg-slate-400",
    }
}

pub(crate) fn get_stroke_from_port_type(port_type: &PortType) -> &'static str {
    match port_type {
        PortType::Output(CommandOutputParameterType::Stdout) => "stroke-slate-300",
        PortType::Output(CommandOutputParameterType::Stderr) => "stroke-slate-200",
        _ => primary_cwl_type(port_type)
            .as_ref()
            .map(cwl_type_to_stroke)
            .unwrap_or("stroke-slate-400"),
    }
}

pub(crate) fn slot_geometry(port_type: &PortType) -> &'static str {
    let is_file_like = matches!(
        primary_cwl_type(port_type),
        Some(CWLType::File) | Some(CWLType::Directory)
    );
    let is_stream = matches!(
        port_type,
        PortType::Output(CommandOutputParameterType::Stdout | CommandOutputParameterType::Stderr)
    );
    if is_file_like || is_stream {
        "rotate-45"
    } else {
        "rounded-lg"
    }
}

pub(crate) fn slot_bg(port_type: &PortType) -> &'static str {
    match port_type {
        PortType::Output(CommandOutputParameterType::Stdout) => "bg-slate-300",
        PortType::Output(CommandOutputParameterType::Stderr) => "bg-slate-200",
        _ => primary_cwl_type(port_type)
            .as_ref()
            .map(cwl_type_to_bg)
            .unwrap_or("bg-slate-400"),
    }
}

pub(crate) fn slot_border(port_type: &PortType) -> &'static str {
    if is_array(port_type) {
        "border border-green-800"
    } else if is_optional(port_type) {
        "border border-red-800"
    } else {
        "border border-1 border-black"
    }
}

File: src/lib/graph/styling.ts
// Ported from studio_legacy/components/graph/styling.rs

import type { NodeKind, PickValue } from "./types";

const BG_BY_TYPE: Record<string, string> = {
  file: "bg-green-400!",
  directory: "bg-blue-400!",
  string: "bg-red-400!",
  boolean: "bg-yellow-400!",
  double: "bg-purple-400!",
  float: "bg-pink-400!",
  long: "bg-cyan-400!",
  int: "bg-teal-400!",
};

function parts(dataType: string): string[] {
  return dataType.split("|").map((p) => p.trim().toLowerCase());
}

// e.g. "File[]" for an array, plain otherwise. Only the outer "| null" split
// is handled here -- a multi-item array like "File | string[]" (rare in
// practice) isn't parsed apart correctly, same limitation the old "array"
// label had, just one level deeper.
function nonNullPart(dataType: string): string {
  return parts(dataType).find((p) => p !== "null") ?? "null";
}

function isOptional(dataType: string): boolean {
  return parts(dataType).includes("null");
}

function isArrayLike(dataType: string): boolean {
  return nonNullPart(dataType).endsWith("[]");
}

// The item type an array port is colored by, e.g. "file" for "File[]" --
// falls back to the bare type for anything that isn't an array.
function primaryType(dataType: string): string {
  const part = nonNullPart(dataType);
  return isArrayLike(dataType) ? part.slice(0, -2) : part;
}

function baseBg(dataType: string): string {
  if (dataType === "stdout") return "bg-slate-300!";
  if (dataType === "stderr") return "bg-slate-200!";
  return BG_BY_TYPE[primaryType(dataType)] ?? "bg-slate-400!";
}

// Solid fill = required; hollow (node-surface color instead of the type
// color, type-colored border instead) = optional. Replaces a colored ring,
// which collided with the type-color fill itself -- a red `string?` port
// got a red ring on a red dot, unreadable regardless of which red it was.
export function portBg(dataType: string): string {
  return isOptional(dataType) ? "bg-bg-surface!" : baseBg(dataType);
}

// Same color as the port itself, as a `stroke` value instead of a `bg-*`
// class -- edges are SVG, not DOM elements Tailwind can target. Reusing
// baseBg() (not portBg()) so an edge out of an optional port still carries
// its real type color rather than the hollow-port "no fill" one; slice off
// both the "bg-" prefix and the trailing "!" it always carries.
export function edgeStrokeColor(dataType: string): string {
  return `var(--color-${baseBg(dataType).slice(3, -1)})`;
}

export function portGeometry(dataType: string): string {
  const t = primaryType(dataType);
  return t === "file" || t === "directory" || dataType === "stdout" || dataType === "stderr"
    ? "rotate-45!"
    : "rounded-full!";
}

const BORDER_BY_TYPE: Record<string, string> = {
  file: "border-green-400!",
  directory: "border-blue-400!",
  string: "border-red-400!",
  boolean: "border-yellow-400!",
  double: "border-purple-400!",
  float: "border-pink-400!",
  long: "border-cyan-400!",
  int: "border-teal-400!",
};

function baseBorderColor(dataType: string): string {
  if (dataType === "stdout") return "border-slate-300!";
  if (dataType === "stderr") return "border-slate-200!";
  return BORDER_BY_TYPE[primaryType(dataType)] ?? "border-slate-400!";
}

// A hollow (optional) port needs its border to carry the type color, since
// the fill no longer does. A solid (required) port keeps the original
// plain, low-key outline -- the "punched through the node" look, not a
// second color signal.
export function portBorder(dataType: string): string {
  return isOptional(dataType) ? `border-2! ${baseBorderColor(dataType)}` : "border! border-black!";
}

// Array gets its own ring, independent of portBorder (already spoken for by
// required/optional) so the two compose instead of colliding on the same
// property -- `outline` doesn't affect layout and isn't touched by
// .svelte-flow__handle, so no `!important` fight to win here.
export function portArrayRing(dataType: string): string {
  return isArrayLike(dataType) ? "outline-2 outline-white outline-offset-1" : "";
}

export function nodeHeaderClass(kind: NodeKind): string {
  switch (kind) {
    case "step":
      return "bg-green-hdr";
    case "input":
      return "bg-blue-hdr";
    case "output":
      return "bg-red-hdr";
  }
}

// Halo ring matching the node body -- the ComfyUI-style "punched through the
// border" look for port dots. box-shadow, not border, so it layers with
// portBorder()'s semantic array/optional ring instead of overwriting it.
export const portRing = "shadow-[0_0_0_2px_var(--color-bg-surface)]";

// pickValue only exists on a multi-source input, so it's shown on the edges
// converging on that port, not on the node -- unlike scatter/when, which are
// step-wide.
export const pickValueEdgeStyle = "stroke-dasharray: 4 3;";
export const pickValueLabelStyle =
  "background: var(--color-bg-panel); border: 1px solid var(--color-border); border-radius: 4px; padding: 1px 6px; font-size: 10px; font-family: 'IBM Plex Mono', monospace; color: var(--color-text-2); white-space: nowrap;";

const PICK_VALUE_LABEL: Record<PickValue, string> = {
  first_non_null: "first non-null",
  the_only_non_null: "only non-null",
  all_non_null: "all non-null",
};

export function pickValueLabel(value: PickValue): string {
  return PICK_VALUE_LABEL[value];
}
